### ISSUE TITLE
[Codegen] Materialize a borrowed slice/array view header once and state its alias facts

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1079]
Baselines: Align `8c8bfbc7a3169e84ecc8415f5149ab8c61afe863` (alignc 0.7.5, LLVM 22.1.8), align-llm
`b72b0b48`, Apple M1, `--profile release`, default `--target-cpu baseline`. Every IR/asm excerpt below is
emitted output.

## Problem

Align lowers a borrowed `slice<T>`/`array<T>` parameter as a *pointer to* its `{ptr,len}` header, and
codegen emits a fresh `load` of the header fields at every use site. Nothing caches the header, no
metadata separates header memory from the element buffer, and a `borrow` header parameter carries neither
`noalias` nor `dereferenceable`. Any loop that also stores through a second view therefore re-reads both
header fields per iteration and LLVM must assume the store invalidates them, so the loop stays scalar.

The 5.3 MB align-llm release image contains **zero floating-point SIMD instructions in any
Align-generated function**, in all three of release / `--target-cpu native` / `--profile fast`
(489 integer DP-SIMD instructions against 20,716 q-register moves, 42:1).

### 1. One load per use, no metadata

`crates/align_codegen_llvm/src/lib.rs:22357-22376` — `borrowed_view_part`:

```rust
        let ptr = self.borrowed_place_ptr(place)?;
        let field = self
            .builder
            .build_struct_gep(slice_struct_type(self.ctx), ptr, index, name)
            .map_err(|e| self.err(e))?;
```

`Rvalue::SliceIndex` (`lib.rs:16095-16116`) reloads the whole aggregate via `self.operand(s)?`. MIR makes
it structural — `hir::Stmt::AssignIndex` (`crates/align_mir/src/lib.rs:6616-6620`) re-lowers
`Rvalue::Load(base)` + `Rvalue::SliceLen` for each element store:

```rust
                    let sv = b.fresh_value(base_ty);
                    b.push(Stmt::Let(sv, Rvalue::Load(*base)));
                    let len = b.fresh_value(i64_ty());
                    b.push(Stmt::Let(len, Rvalue::SliceLen(Operand::Value(sv))));
                    emit_bounds_check(b, &idx, Operand::Value(len));
```

`!range`/`!nonnull`/`!align`/`!tbaa` on loads: **0** across the 746,869-line whole-program optimized IR.
`mark_borrow_param_contracts_at` (`crates/align_codegen_llvm/src/lib.rs:7216-7226`) gives
`ParamMode::Borrow` only `nonnull` + `captures(none)` + `readonly` and `continue`s on every other mode, so
`borrow mut` (passed as `{ptr,ptr}` = header pointer + drop-flag pointer) gets no attributes at all:

```rust
    for (index, mode) in modes.iter().copied().enumerate() {
        if mode != align_ast::ParamMode::Borrow {
            continue;
        }
```

### 2. The smallest possible in-place SIMD kernel stays scalar

```align
fn scale_only(borrow mut out: array<f32>, k: f32) {
  mut i := 0
  loop { if i >= out.len() { break }  out[i] = out[i] * k  i = i + 1 }
}
```

`--stage optimized`, release:

```llvm
bb9:
  %load1.unpack = load ptr, ptr %borrow.mut.ptr, align 8   ; data ptr RELOADED every iteration
  %slcidx = getelementptr inbounds nuw float, ptr %load1.unpack, i64 %_2.023
  %slcload = load float, ptr %slcidx, align 4
  %fmul = fmul float %1, %slcload
  store float %fmul, ptr %slcidx, align 4
  %load1.unpack19 = load i64, ptr %load1.elt18, align 8    ; len RELOADED every iteration
  %ge.not = icmp slt i64 %add, %load1.unpack19
```

`--profile fast --target-cpu native` gives the same 18-instruction scalar loop, zero vector instructions.
Hoisting `count := out.len()` in source does not help: the bounds check reloads `len` anyway. There is no
user-side workaround.

Measured against a clang `-O2` C twin (4096 f32 × 60,000 reps, 3 interleaved runs). Both fixes were
applied to the emitted *raw* IR only and re-run through the identical `default<O2>` pipeline with
`llvm@22/bin/opt`; `vector.body` went 0 → 5 in each case:

| kernel | Align release / native | patched | clang `-O2` C twin |
|---|---|---|---|
| `out[i] = out[i] * k` | 0.1386 / 0.1435 / 0.1444 s | **0.0170 s** (2-node TBAA tree) | 0.0150 s |
| `out[i] = out[i] + a[i]*k` | 0.2565 / 0.2581 / 0.2716 s | **0.0225 s** (header hoisted to preheader) | 0.0199–0.0267 s |

### 3. `noalias` on a read-only `borrow` header hoists `len` — measured on real client code

Taking the emitted body of `decode_step$top_k` unchanged and adding only `noalias` to its read-only
`borrow` slice header parameter, then re-running `default<O2>`:

| variant | in-loop `len` loads | in-loop data-`ptr` loads | `vector.body` |
|---|---|---|---|
| as emitted | 3 | 2 | 0 |
| `+ noalias` on the `borrow` header param | **1** | 2 | 0 |

A real, cheap win for the `len` field — and not sufficient on its own.

### 4. The `borrow` vs `out` asymmetry is what defeats LLVM

A read-only `borrow` passes a *pointer* to the header; an `out` slice passes `{ptr,i64}` *by value*.
Emitted signature of the "route ggml's byte logits into a typed `f32` destination" kernel:

```llvm
define void @bytes_to_f32_out(ptr nonnull readonly captures(none) %0, { ptr, i64 } %1)
```

`%0` has no `noalias`, so the `store` through `%1` may alias the header and both fields are reloaded per
element — 10 instructions per single f32 copied:

```
2c: ldr  x2, [x0, #0x8]      ; LEN reloaded
30: add  x9, x8, #0x4
34: cmp  x9, x2
38: b.gt <range trap>        ; 1 surviving in-loop check
3c: ldr  x11, [x0]           ; PTR reloaded
40: subs x10, x10, #0x1
44: ldr  s0, [x11, x8]
48: str  s0, [x1, x8]
4c: mov  x8, x9
50: b.ne 0x2c
```

### 5. Conjunctive with the bounds-check work

Four variants of that same function through the identical `default<O2>`:

| variant | `vector.body` | `<4 x float>` ops |
|---|---|---|
| A — as emitted | 0 | 0 |
| C — in-loop checks hoisted, no `noalias` | **0** | 0 |
| D — `noalias` on the header, checks as emitted | **0** | 0 |
| B — both | **2** (main + epilogue) | **10** |

Neither factor alone produces a single vector instruction. Measured on 152,064 elements, variants A/C/D
all sit at **59.4 µs/call** and variant B at **12.6 µs/call (4.7×)**; an earlier run of the same pair
measured 65.7 → 11.4 µs, i.e. the same effect. Fixing one half yields nothing, so this issue and
#1081 should be scheduled as one unit.

### 6. LLVM's own diagnosis on 26,957 lines of hot align-llm source

```
$ grep -oh "because [a-z][^\"]*" explain/*.verbose.txt | sort | uniq -c | sort -rn
  14803 because its definition is unavailable
  12185 because the loop may invalidate its value      <- "failed to move load with loop-invariant address"
    311 because load is conditionally executed          <- "failed to hoist load with loop-invariant address"
```

Top LICM-failure sites (`file:line:col`, count): `layer_forward.align:1316:10` (284), `:1315:3` (276),
`:1740:7` (236), `:1745:9` (222), `:1348:18` (214), `moe_layer_forward.align:1515:10` (122),
`:1962:7` (118), `runtime_qwen_load.align:92:49` (106). 560 of the 12,185 land on one three-line
helper, `align-llm/src/layer_forward.align:1314-1317`:

```align
fn column_at(borrow column: array<i64>, index: i64) -> i64 {
  if index < 0 || index >= column.len() { return 0 }
  return column[index]
}
```

The 311 conditional-hoist failures spread over the two-slice-index shape: `layer_forward.align:1329:5`
(62), `model_forward.align:1175:5` (31), `:1165:5` (31), `moe_layer_forward.align:1596:5` (30),
`layer_olmoe.align:1339:9` (15).

### 7. Scale in the real client

`grep -c 'borrow mut' src/*.align` = **811 lines** (856 occurrences); indexed element-assignment
sites = **222**. The hottest concrete instance is `align-llm/src/decode_step.align:636-660`
(`stage_past_k`), whose inlined inner loop in the shipped binary is nine instructions, two bounds
branches, **two header reloads** and one opaque call per 4 bytes moved; the disassembly is in
#1081 §4, where the same loop is the bounds-check evidence.

## Root cause

1. `borrowed_view_part` has no per-function or per-loop cache: one `load` per syntactic use.
2. No aliasing metadata distinguishes the `{ptr,len}` header slot from the element buffer, although in
   Align's own lowering these are structurally distinct memories (header in a parameter/local slot,
   elements in a separately allocated buffer).
3. `borrow`/`borrow mut` header parameters carry no `noalias`/`dereferenceable`; `borrow mut` carries no
   parameter attributes whatsoever.

## Proposed solution

**(a) TBAA node separating the view header from element storage** — a two-node tree
(`align.view.header` vs `align.elem.<T>`). This states a fact about Align's own representation, not one
minted from source spelling, so it is outside the `docs/impl/63-codegen-performance-audit.md:218` /
`64-composed-byte-optimization-plan.md:186` prohibition. TBAA has **zero** mentions in all of `docs/`.

**(b) Materialize the header once per function (or per loop preheader) instead of per use** — needs no new
LLVM assertion; this is the variant that recovered 9.4× on AXPY. Mechanically a cache keyed on
`(BorrowedPlace, field)` in `borrowed_view_part`, plus reusing an existing `SliceLen` in the MIR
`AssignIndex` lowering instead of re-emitting it.

**(c) `noalias dereferenceable(16) align 8` on `borrow` slice/array header parameters**, and revisit the
`continue` at `crates/align_codegen_llvm/src/lib.rs:7224-7226` so `borrow mut` gets `nonnull`,
`dereferenceable(16)` and `align 8` on its header pointer. Align's exclusivity rules already guarantee
what `noalias` asserts about a `borrow` header slot.

What does **not** work, so it is not re-attempted: `noalias` + `dereferenceable(16)` + `align 8` on the
`borrow` header alone leaves the loop scalar (`vector.body` 0; 1 of 4 header loads removed), and so does
adding `memory(argmem: readwrite)` to the runtime calls. (a) or (b) is required.

## Acceptance criteria

- `scale_only` emits one header load in the preheader and a `<4 x float>` `vector.body` at
  `--profile release --target-cpu baseline`; the two-slice AXPY shape gets a vector main loop. Both land
  within ~1.5× of the clang `-O2` C twin on the same host.
- On the real `decode_step$top_k` body, in-loop `len` loads go 3 → 1 and data-pointer loads 2 → 1.
- align-llm's `because the loop may invalidate its value` remark count drops by at least an order of
  magnitude from 12,185; `column_at`'s 560 remarks go to 0.
- A codegen owner test pins the metadata/attribute shape;
  `crates/align_driver/tests/vectorize_shapes.rs` passes unchanged.
- `bytes_to_f32_out` vectorizes only with both this change and the bounds-check change; pin that
  conjunction in the test.

## Relation to existing issues/plans

This is one of three parts of a single **loop-facts contract** — what an Align loop tells LLVM about its
own data. The three should be designed and landed together, not as separate patches:

1. **this issue** — header materialization and alias facts (`{ptr,len}` is loaded once, and header
   memory is known distinct from element memory);
2. **len facts** (`!range` on length loads) — *"[Codegen] Attach `!range` to slice/array length
   loads so a length is known non-negative"* (#1080);
3. **check shape** (one unsigned compare, monotone-induction checks in the preheader) — *"[MIR] Fuse
   the bounds check into one unsigned compare and eliminate monotone-induction in-loop checks"*
   (#1081).

§5 shows why: two of the three in isolation produce **zero** vector instructions on a real kernel;
the conjunction produces a `<4 x float>` main loop. A partial fix is not measurable, so the contract
is the unit of work.

- Not owned by any plan. `docs/impl/07-roadmap.md:2350-2362` deferred per-argument attribute work because
  *"LLVM's FunctionAttrs already infers ... with zero codegen help"*; the measurements above are the
  counter-example for the `borrow`-header-plus-store shape, and (a)/(b) are codegen-form changes, not
  attribute claims.
- Conjunctive with the bounds-check fusion / MIR BCE issue, #1081 (§5).
- Overlaps the mechanism behind #1064 (logits argmax not vectorized), but is more general: #1064's
  `slice<u8>` → typed-`f32` view ask is the *access* half, this is the *aliasing* half.
- `docs/open-questions.md:4187-4189` rejects `llvm.assume` and loop-metadata overrides as a general
  policy. Proposals (a)–(c) use neither.

## Impact framing

Honest scope: in a real Metal generation profile Align-compiled code is ~0.1 % of sampled CPU during
GPU decode, so this is not a tokens/s claim today. It reaches the CPU-side per-token kernels Align
owns (greedy logits selection **242 µs/token** over 151,936 logits, sampler **267 µs/token** over
152,064 — `scripts/bench-runtime-greedy` / `-sampler` on this host); the
tokenizer path, which is **74.5 % Align code**; model load, KV-plane verification and digest paths;
and the future in-language CPU inference path. Independent of any workload, 12,185 self-reported LICM
failures and zero FP SIMD instructions in a 5.3 MB numeric binary is a code-quality statement.
